### ISSUE TITLE
Crash when calling generic method with same-type requirement on constrained existential

### ISSUE THREAD

**Issue by hisekaldma** [apple/swift#61309]
**Describe the bug**
When using a constrained existential `any P<Int?>` of a protocol `P` with a primary associated type `T`, and the protocol has a generic method with a same-type requirement that constrains the generic parameter based on the primary associated type `func<U>(_: U) where T == U?`, calling the method on the existential crashes.

**Steps To Reproduce**
Paste the following into a .swift file and run it:
```swift
protocol P<T> {
    associatedtype T
    func foo<U>(_: U) where T == U?
}

struct PImpl<T>: P {
    var t: T
    func foo<U>(_ u: U) where T == U? {
        print(u)
    }
}

let p: any P<Int?> = PImpl(t: 0)
p.foo(1)
```

**Expected behavior**
The code should either a) not compile, or b) compile and print 1. Currently, it compiles but crashes when run.

**Environment (please fill out the following information)**
 - OS: macOS 12.6
 - Xcode Version/Tag/Branch: Xcode 14.0

**Comment by hisekaldma** [comment]
@slavapestov Do you know if this is supposed to work?

**Comment by slavapestov** [comment]
I believe this should work. I'll take a look.

**Comment by slavapestov** [comment]
No, I'm wrong. This should be rejected for the same reason that we reject

```swift
protocol P<T> {
    associatedtype T
    func foo(_: Int) where T == Int?
}
```

You can't have a protocol requirement that doesn't apply to all possible substitutions of `Self` (imagine if your conforming type witnessed `T` with something that could never be an Optional. What would be signature of the protocol requirement be?).

**Comment by slavapestov** [comment]
This is indeed unsound, but we now warn about it:
```
diag.swift:3:10: warning: instance method requirement 'foo' cannot add constraint 'Self.T == U?' on 'Self'; this will be an error in a future Swift language mode
 1 | protocol P<T> {
 2 |     associatedtype T
 3 |     func foo<U>(_: U) where T == U?
   |          `- warning: instance method requirement 'foo' cannot add constraint 'Self.T == U?' on 'Self'; this will be an error in a future Swift language mode
 4 | }
 5 | 
```
